validator/remote/src/main/java/tech/pegasys/teku/validator/remote/EventSourceBeaconChainEventAdapter.java

```diff
@@ -14,78 +14,188 @@
+    return SafeFuture.fromRunnable(
+            () -> {
+              primaryEventSource.close();
+              closeFailoverEventSource();
+              cancelScheduledPingingTask();
+            })
+        .thenCompose(__ -> timeBasedEventAdapter.stop())
+        .thenRun(runningLatch::countDown);
+  }
+
+  private EventSource createPrimaryEventSource() {
+    return defaultEventSourceBuilder(primaryApiEndpoint)
+        .connectionErrorHandler(
+            throwable -> {
+              if (failoverApiEndpoints.isEmpty()) {
+                return Action.PROCEED;
+              }
+              if (!PingUtils.hostIsReachable(primaryApiEndpoint)) {
+                // connect to a failover Beacon node in case the primary Beacon node is not
+                // reachable (offline)
+                final Optional<HttpUrl> maybeFailover = findAvailableFailover();
+                final Optional<URI> maybeFailoverUri = maybeFailover.map(HttpUrl::uri);
+                VALIDATOR_LOGGER.beaconNodeIsNotReachableForEventStreaming(
+                    primaryApiEndpoint.uri(), maybeFailoverUri);
+                if (maybeFailover.isPresent()) {
+                  startFailoverEventSource(maybeFailover.get());
+                  return Action.SHUTDOWN;
+                }
+              }
+              return Action.PROCEED;
+            })
+        .build();
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void startFailoverEventSource(final HttpUrl failoverApiEndpoint) {
+    asyncRunner
+        .runAsync(
+            () -> {
+              // wait until the primary Beacon node event source is shut down
+              SafeFuture.asyncDoWhile(
+                  () ->
+                      SafeFuture.completedFuture(
+                          primaryEventSource.getState() != ReadyState.SHUTDOWN));
+            })
+        .thenRun(
+            () -> {
+              final EventSource failoverEventSource =
+                  defaultEventSourceBuilder(failoverApiEndpoint).build();
+              LOG.info(
+                  "Connecting to a failover Beacon Node {} for event streaming.",
+                  failoverApiEndpoint);
+              failoverEventSource.start();
+              maybeFailoverEventSource = Optional.of(failoverEventSource);
+              schedulePingingOfPrimaryBeaconNode();
+            });
+  }
+
+  private EventSource.Builder defaultEventSourceBuilder(final HttpUrl apiEndpoint) {
+    return new EventSource.Builder(eventSourceHandler, createHeadEventSourceUrl(apiEndpoint))
+        .maxReconnectTime(MAX_RECONNECT_TIME)
+        .client(okHttpClient);
+  }
+
+  private void schedulePingingOfPrimaryBeaconNode() {
+    final Cancellable cancellable =
+        asyncRunner.runWithFixedDelay(
+            () -> {
+              final boolean primaryBeaconNodeIsReachable =
+                  PingUtils.hostIsReachable(primaryApiEndpoint);
+              if (primaryBeaconNodeIsReachable) {
```

Comment:
I suspect here we need to just keep trying to connect to the primary event stream and then when it reconnects drop the fail over connection. I suspect the key to making this work well is going to be a wrapper around `EventSourceHandler` that knows which event source it's getting events from and can handle the onOpen and onClosed events to trigger starting or stopping a fail-over connection. I suspect that would also let us leave reconnecting to the primary to the built-in retry capability of `EventSource` which has some nice back-off handling and so on.